Add omitempty support in INSERT statements

letFunny opened the pull request:
When `omitempty` tag is present in a struct field and the field is set to the zero value, it will not be included in the INSERT. This way we can support columns which auto increment.

## How does it work internally?
We prepare the statement in the usual way and when supplied the zero value for the field (i.e. 0 for an int that serves as an id) and the field was annotated with `omitempty`, we instead send `nil`.

## Is sending NULL supported in every DB?
No, it is supported in MySQL and in SQLite but not in Postgres. However, the sql Go library does not support sending any other form of default. That is why we have decided to use NULL and in the future we can look for another solution if the need arises.

For context, several DBs instead use `DEFAULT` for this use case, the problem is that SQLite does not.

## Can we just omit the field in the query altogether?
Because of how SQLair works, we prepare a statement before getting the values, meaning we have we have to know the number of fields prior to knowing if we are going to omit any, which makes it impossible to leave out a field directly in Prepare.

## Alternatives
If we want to support this use case in all the different DBs in the future we could do the following:
* [PREFERRED] Re-prepare the statement each time we want to set the field with the default being omitting it. The assumption here is that the majority of use cases would be excluding the field, and the performance penalty would be small.
* Open an issue in the std sql library for them to support sending a default value in a prepared statement in a database agnostic way.

letFunny committed:
Add omitempty support in INSERT statements

When omitempty tag is present in a struct field and the field is set to
the zero value, it will not be included in the INSERT. This way we can
support columns which auto increment.

Aflynn50 commented on example_omitempty_test.go:
```diff
@@ -0,0 +1,59 @@
+	dave := Person{Fullname: "James", PostalCode: 4500}
+	allPeople := []Person{fred, mark, mary, dave}
+
+	sqldb, err := sql.Open("sqlite3", ":memory:")
```
It would be best to use the full string found in `package.go` tests here rather then ":memory:" so that we are consistent across the codebase.

letFunny replied:
Synced offline, we will use `file:exampleOmitTest.db?cache=shared&mode=memory` to prevent errors. When go database.sql internally spawns a new connection, `:memory:` will create a totally new db resulting in errors such as "no table found".

Aflynn50 commented on example_omitempty_test.go:
```diff
@@ -0,0 +1,59 @@
+	"github.com/canonical/sqlair"
+)
+
+func Example() {
```
Might be best to call this `Example_omitempty` so that its clear this is not a whole package example. It should still be recognised if I am interpreting the [blog post](https://go.dev/blog/examples) correctly

letFunny replied:
I will include it to further differentiate that this example is about `omitempty` but it is still counted as a whole package example. I tested it locally using `godoc` with the name change.

Aflynn50 commented on internal/expr/parts.go:
```diff
@@ -43,6 +43,10 @@ func (p *inputPart) String() string {
 
 func (p *inputPart) part() {}
 
+func (p *inputPart) isInsert() bool {
```
I'm not really sure this extra function in necessary for this PR, we can change it in a general improvements PR in the future, but right now I think we can get away with maybe just adding comments where it is unclear.

letFunny replied:
I had to add a check to see if a given inputPart was part of an insert statement. Although I agree with you it is not strictly related to this PR, I think the effort of copying the check and adding a comment is the same as this small refactor which will make it clearer for all usecases.

Aflynn50 replied:
Fair enough.

Aflynn50 commented on internal/expr/typeinfo.go:
```diff
@@ -115,6 +115,18 @@ func (mi *mapInfo) getAllMembers() ([]typeMember, error) {
 	return nil, fmt.Errorf(`map type %q cannot be used with asterisk`, mi.mapType.Name())
 }
 
+type sqlContext int
```
For now I think we could just have a boolean here, that indicates if it is an insert statement. If we need more cases in the future we could add something like this.

letFunny replied:
Changed

Aflynn50 commented on package_test.go:
```diff
@@ -1321,3 +1321,63 @@ func (s *PackageSuite) TestIterMethodOrder(c *C) {
 	err = iter.Close()
 	c.Assert(err, IsNil)
 }
+
```
These could all be defined inside the test.

Aflynn50 commented on package_test.go:
```diff
@@ -1321,3 +1321,63 @@ func (s *PackageSuite) TestIterMethodOrder(c *C) {
+	c.Assert(err, IsNil)
+	err = db.Query(nil, createPerson).Run()
+	c.Assert(err, IsNil)
+	defer dropTables(c, db, "personAutoId")
```
Should that string be "personAutoID"? Seems like SQLite is case insensitive, but still worth changing.

letFunny replied:
Nice catch.

Aflynn50 commented on package_test.go:
```diff
@@ -1321,3 +1321,63 @@ func (s *PackageSuite) TestIterMethodOrder(c *C) {
+	for i, p := range outPeople {
+		c.Assert(p.Fullname, Equals, allPeopleAutoId[i].Fullname)
+		c.Assert(p.PostalCode, Equals, allPeopleAutoId[i].PostalCode)
+		// Numeration starts at 1.
```
s/Numeration/Enumeration

Aflynn50 commented on package_test.go:
```diff
@@ -1321,3 +1321,63 @@ func (s *PackageSuite) TestIterMethodOrder(c *C) {
+		// Numeration starts at 1.
+		c.Assert(p.ID, Equals, i+1)
+	}
+
```
Same as previous block

Aflynn50 commented on package_test.go:
```diff
@@ -1321,3 +1321,63 @@ func (s *PackageSuite) TestIterMethodOrder(c *C) {
+		err := db.Query(nil, insertPerson, p).Run()
+		c.Assert(err, IsNil)
+	}
+
```
Maybe put the comment explaining what this block does above the block.

Aflynn50 reviewed (commented):
Looks very good! Nice and clean code.

I do have to admit I'm not quite sure about the `inputTypeMember` struct, to me it feels a bit weird having the clutter of the whole extra struct only for that information. I think it would probably be cleaner to set `omitEmpty` in the plain `structField` struct and have some clear comments to explain what is going on. I can probably be convinced otherwise though if you are against the idea.

letFunny commented:
> Looks very good! Nice and clean code.
> 
> I do have to admit I'm not quite sure about the `inputTypeMember` struct, to me it feels a bit weird having the clutter of the whole extra struct only for that information. I think it would probably be cleaner to set `omitEmpty` in the plain `structField` struct and have some clear comments to explain what is going on. I can probably be convinced otherwise though if you are against the idea.

My reasoning was the following:
* I did not want to have to change something on an already instantiated `structField` as that would mean having mutable state that would not be obvious (i.e. having to call a function prior to filling the data with no hint in the type system). 
* That we might want to extend this notion of context to other inputs apart from `structField` for things like better error messages, that is why I did not want to include a flag in `structField` directly.

but I can also be convinced otherwise :).

Another idea, what do you think about embedding `typeMember` in the struct which will make the code much more lightweight?

Aflynn50 commented:
> > Looks very good! Nice and clean code.
> > I do have to admit I'm not quite sure about the `inputTypeMember` struct, to me it feels a bit weird having the clutter of the whole extra struct only for that information. I think it would probably be cleaner to set `omitEmpty` in the plain `structField` struct and have some clear comments to explain what is going on. I can probably be convinced otherwise though if you are against the idea.
> 
> My reasoning was the following:
> 
>     * I did not want to have to change something on an already instantiated `structField` as that would mean having mutable state that would not be obvious (i.e. having to call a function prior to filling the data with no hint in the type system).
> 
>     * That we might want to extend this notion of context to other inputs apart from `structField` for things like better error messages, that is why I did not want to include a flag in `structField` directly.
> 
> 
> but I can also be convinced otherwise :).
> 
> Another idea, what do you think about embedding `typeMember` in the struct which will make the code much more lightweight?

The embedding sounds good!

letFunny committed:
Address PR comments

Embedding struct instead of adding a new field, renaming identifiers and
some refactoring.

letFunny committed:
Change var declaration to :=

letFunny committed:
Merge branch 'new-add-insert-values' of github.com:Aflynn50/sqlair into new-add-insert-values
